src/dsp/rx_fft.cpp

```diff
@@ -219,6 +220,8 @@ void rx_fft_c::set_window_type(int wintype)
 
     d_window.clear();
     d_window = gr::fft::window::build((gr::fft::window::win_type)d_wintype, d_fftsize, 6.76);
+    volk_32f_accumulator_s32f(&tmp, d_window.data(), d_fftsize);
+    volk_32f_s32f_normalize(d_window.data(), tmp / float(d_fftsize), d_fftsize);
```

Comment:
GNU Radio 'normalization', converted from lambdas and std algorithms to libvolk may look like this:
```
    volk_32f_x2_dot_prod_32f(&tmp, d_window.data(), d_window.data(), d_fftsize);
    volk_32f_s32f_normalize(d_window.data(), std::sqrt(tmp / float(d_fftsize)), d_fftsize);
```
2 lines instead of 14...
Libvolk looks much less cryptic to me.

Test results:
![flattop](https://user-images.githubusercontent.com/17007837/219210760-41ece454-64d3-44c2-b2d4-fb8f653f84cf.png)
![kaiser](https://user-images.githubusercontent.com/17007837/219210771-38919af6-cb06-4897-bc27-6e6ba66801db.png)
![rect](https://user-images.githubusercontent.com/17007837/219210778-4c0a478e-193c-4a8a-8171-c662c5e33fae.png)

The rectangular window gives correct level, other windows are giving wrong level...
So, GNU Radio 'normalization' does not fit well here.

Reply:
Why do you you say the level is wrong? The bin powers appear to add to equal values when the RMS calculation is used, but the original calculation shows differing amounts of power between the windows.

Reply:
What is the rationale for multiplying by `d_fftsize` (or its square root, in the RMS calculation)?

Reply:
> In case of strong CW signal, say ~30db over the noise floor or more, there will be negligible amount of leakage from adjacent bins

I don't think that's the case. Spectral leakage is a property of the window function, and will occur even for CW signals.

> Scaling the window to AVG=1.0 makes all windows show almost equal level for the same CW signal. And this level matches the S-meter measurement.

I don't see any mathematical justification for this. The total power displayed will be higher than reality if this calculation is used, and it doesn't actually result in the centre bin being equal between windows.

Even if some compensation was applied to make CW signals appear to have the same power in the centre bin, there would still be scalloping loss if the signal was not exactly centred in a bin.

I think RMS normalization (as GNU Radio does) is the way to go, since the total displayed power across all bins will be correct regardless of the window function.

Reply:
> it doesn't actually result in the centre bin being equal between windows

Hmmm. I can't confirm this. My tests are showing that the centre bin is roughly equal (up to 3db less due to scalloping loss when the signal is exactly between bins) between windows and it is always equal in case of flattop window as it minimizes the scalloping loss.

RMS normalization does not play well with partially negative windows (ex. flattop).

> since the total displayed power across all bins will be correct regardless of the window function.

The FFT is calculated over voltage, not power. The result is converted to power here https://github.com/gqrx-sdr/gqrx/blob/master/src/applications/gqrx/mainwindow.cpp#L1379-L1383
So, applying 'power-normalized' window results in power-normalization applied twice. Is it correct? Or, maybe, I'm wrong.

Reply:
After looking into this some more, I think we are both right. :smile:

When applying a window function, you can apply either an amplitude correction factor (1 / mean(w)), or an energy correction factor (1 / rms(w)):

https://dsp.stackexchange.com/questions/85277/what-is-the-energycorrection-or-amplitude-correction-for-a-tukey-window

I guess Gqrx will have to decide which of those to use. It might be worth checking what other programs do when displaying an FFT.

Since the FFT displays the signal power, it may make sense to normalize the energy/power of the window, but I don't know whether this is the convention.

Reply:
And on the other hand, amplitude normalization should result in a CW signal having its peak at the right amplitude, so there's also an argument for that.

Reply:
> I think we are both right.

OK. I'll implement all 3 cases and update the PR.
Is this GUI selector fine?
![window scaling options](https://user-images.githubusercontent.com/17007837/219436004-36faca2c-b8c8-4b92-a050-e2409ba99abf.png)
![window scaling](https://user-images.githubusercontent.com/17007837/219436010-a3788750-b2de-43ef-96b8-05b3f5fcf10d.png)

Reply:
I don't think it's necessary to make it configurable.

I think you're correct that users would probably care the most about CW peaks being at the correct magnitude, so we could just go with amplitude normalization.

Reply:
OK